server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -58,8 +58,10 @@
  * Listens for changes in the number of data nodes and immediately submits a
  * ClusterInfoUpdateJob if a node has been added.
  *
- * Every time the timer runs, gathers information about the disk usage and
- * shard sizes across the cluster.
+ * Every time the timer runs, if <code>cluster.routing.allocation.disk.threshold_enabled</code>
+ * is enabled, gathers information about the disk usage and shard sizes across the cluster,
+ * computes a new cluster info and notifies the registered listeners. If disk threshold
+ * monitoring is disabled, listeners are called with an empty cluster info.
  */
 public class InternalClusterInfoService implements ClusterInfoService, ClusterStateListener {
```

Comment:
This class is strangely complicated, it took quite some effort to work this change through. I'm probably partly to blame here. I opened https://github.com/elastic/elasticsearch/issues/88587.